packages/client/src/task-queue-client.ts

```diff
@@ -0,0 +1,271 @@
+ *   - `NewWorkflows`: The Build Id might be used by new workflows
+ *   - `ExistingWorkflows` The Build Id might be used by open workflows and/or closed workflows.
+ *   - `OpenWorkflows` The Build Id might be used by open workflows
+ *   - `ClosedWorkflows` The Build Id might be used by closed workflows
+ */
+export type ReachabilityType = 'NewWorkflows' | 'ExistingWorkflows' | 'OpenWorkflows' | 'ClosedWorkflows';
+
+interface BaseReachabilityOptions {
+  /**
+   * A list of build ids to query the reachability of. Currently, at least one Build Id must be
+   * specified, but this restriction may be lifted in the future.
+   */
+  buildIds: string[];
+  /**
+   *  A list of task queues with Build Ids defined on them that the request is
+   *  concerned with.
+   */
+  taskQueues?: string[];
+  /** The kind of reachability this request is concerned with. */
+  reachability?: ReachabilityType;
+}
+
+export interface ReachabilityResponse {
+  /** Maps Build Ids to their reachability information. */
+  buildIdReachability: Record<string, BuildIdReachability>;
+}
+
+type ReachabilityTypeResponse = ReachabilityType | 'NotFetched';
+
+export interface BuildIdReachability {
+  /**
+   *  Maps Task Queue names to how the Build Id may be reachable from them. If they are not
+   *  reachable, the map value will be an empty array.
+   */
+  taskQueueReachability: Record<string, ReachabilityTypeResponse[]>;
+}
+
+function reachabilityTypeToProto(type: ReachabilityType | undefined | null): temporal.api.enums.v1.TaskReachability {
+  switch (type) {
+    case null:
+    case undefined:
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_UNSPECIFIED;
+    case 'NewWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_NEW_WORKFLOWS;
+    case 'ExistingWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_EXISTING_WORKFLOWS;
+    case 'OpenWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_OPEN_WORKFLOWS;
+    case 'ClosedWorkflows':
+      return temporal.api.enums.v1.TaskReachability.TASK_REACHABILITY_CLOSED_WORKFLOWS;
+    default:
+      assertNever(type);
+  }
+}
+
+export function reachabilityResponseFromProto(resp: GetWorkerTaskReachabilityResponse): ReachabilityResponse {
+  const buildIdReachability: Record<string, BuildIdReachability> = {};
+  for (const bir of resp.buildIdReachability) {
+    const taskQueueReachability: Record<string, ReachabilityTypeResponse[]> = {};
+    if (bir.taskQueueReachability != null) {
+      for (const tqr of bir.taskQueueReachability) {
+        if (tqr.taskQueue == null) {
+          continue;
+        }
+        if (tqr.reachability == null) {
+          taskQueueReachability[tqr.taskQueue] = [];
+          continue;
+        }
+        taskQueueReachability[tqr.taskQueue] = tqr.reachability.map(reachabilityTypeFromProto);
+      }
+    }
+    buildIdReachability[bir.buildId ?? ''] = { taskQueueReachability };
```

Comment:
I mentioned this before, using empty strings for keys is weird, let's make a symbol for the unversioned build ID (and please export it).

```ts
export const UnversionedBuildId = Symbol.for('__temporal_unversionedBuildId')
```

This function will return 
```
 { buildIdReachabilty: Record<string | UnversionedBuildId, BuildIdReachability> }
```

And in the request you'll accept
```
buildIds: Array<string | UnversionedBuildId>
```